### PR TITLE
Let TouchIcon use updated paths

### PR DESCRIPTION
**Pull request by linc**
Fixes Touch Icons not being able to be uploaded on cloud with CDN.

### CODE REVIEW

**Comment by tburry** [comment]
This is not correct.

* You must generate a new random filename when uploading a new image or else CDN cache issues will arise.
* I really want stuff like this to save into a subfolder such as `touch-icons`.

**Comment by linc** [comment]
Changes:

* Use standard-format config name (matching Banner settings)
* Use 'banner' subfolder (where I'd like to move the rest of the Banner images ultimately too).
* Add backward compatibility with current setup via Structure.

Ultimately I'd like this to be on the Banner page with the rest of the settings, hence some of the choices here.

**Comment by beckyvb** [comment]
Looks good to me!

File: plugins/TouchIcon/class.touchicon.plugin.php [hunk 1]
```diff
@@ -43,33 +60,45 @@ public function SettingsController_TouchIcon_Create($Sender) {
...
+      $Sender->Render($this->GetView('touchicon.php'));
+   }
+
+   /**
+    * Get path to icon.
+    *
+    * @return string Path to icon
+    */
+   public function getIconPath() {
+      $Icon = C('Garden.TouchIcon') ? Gdn_Upload::Url(C('Garden.TouchIcon')) : self::DEFAULT_PATH;
```
Comment (tburry) [inline plugins/TouchIcon/class.touchicon.plugin.php]: You have one case returning a url and the other returning a path. Won't that fail in subfolders?

File: plugins/TouchIcon/class.touchicon.plugin.php [hunk 2]
```diff
@@ -77,10 +106,8 @@ public function SettingsController_TouchIcon_Create($Sender) {
     * @access public
     */
    public function UtilityController_ShowTouchIcon_Create($Sender) {
-      $DefaultPath = 'plugins/TouchIcon/design/default.png';
-      $IconPath = Gdn_Upload::Url('TouchIcon/apple-touch-icon.png');
-      $Redirect = (C('Plugins.TouchIcon.Uploaded')) ? $IconPath : $DefaultPath;
+      $Redirect = $this->getIconPath();
```
Comment (tburry) [inline plugins/TouchIcon/class.touchicon.plugin.php]: Usually endpoints like this need to be 302s.

File: plugins/TouchIcon/class.touchicon.plugin.php [hunk 3]
```diff
@@ -25,14 +31,25 @@ public function Setup() {
...
    }
-   
+
+   /**
+    * Updates.
+    */
+   public function Structure() {
+      // Backwards compatibility with v1.
+      if (C('Plugins.TouchIcon.Uploaded')) {
+         SaveToConfig('Garden.TouchIcon', PATH_ROOT.'/uploads/TouchIcon/apple-touch-icon.png');
```
Comment (tburry) [inline plugins/TouchIcon/class.touchicon.plugin.php]: This seems wrong. You are mixing an absolute path with an uploaded path.

File: plugins/TouchIcon/class.touchicon.plugin.php [hunk 4]
```diff
@@ -43,33 +60,45 @@ public function SettingsController_TouchIcon_Create($Sender) {
...
                $Upload->SaveImageAs(
                   $TmpImage,
-                  PATH_ROOT.'/uploads/TouchIcon/apple-touch-icon.png',
+                  $TouchIconPath,
                   114,
                   114,
                   array('OutputType' => 'png', 'ImageQuality' => '8')
                );
-               SaveToConfig('Plugins.TouchIcon.Uploaded', TRUE);
+               SaveToConfig('Garden.TouchIcon', $TouchIconPath);
```
Comment (tburry) [inline plugins/TouchIcon/class.touchicon.plugin.php]: The `$Upload->SaveImageAs()` method returns an array with a `SavePath` key (I think). That should be used in the config instead.

File: plugins/TouchIcon/class.touchicon.plugin.php [hunk 5]
```diff
@@ -43,33 +60,45 @@ public function SettingsController_TouchIcon_Create($Sender) {
...
             }
          } catch (Exception $ex) {
             $Sender->Form->AddError($ex->getMessage());
          }
-         
+
          $Sender->InformMessage(T('TouchIconSaved', "Your icon has been saved."));
```
Comment (tburry) [inline plugins/TouchIcon/class.touchicon.plugin.php]: Just use "Your icon has been saved." as the translation code.

**Pull request merged**